src/libasr/ASR.asdl

```diff
@@ -278,6 +281,8 @@ expr
 
     | BitCast(expr source, expr mold, expr? size, ttype type, expr? value)
     | DerivedRef(expr v, symbol m, ttype type, expr? value)
+    | EnumName(symbol v, ttype enum_type, ttype type, expr? value)
```

Comment:
Isn't `enum_type` the same as `symbol_type(v)`?

Reply:
No. `enum_type` is the original `EnumType` wrapped under `Enum` node. Type of `v` is same as member variable's type. `enum_type` is needed to fetch the global array in C and LLVM backend to lookup the name and values for non-integers.

Reply:
`enum_type` seems to be the same as `symbol_symtab(v)->asr_owner`.

Reply:
Its not always the case. For the following test for example,

https://github.com/lcompilers/lpython/blob/2d52f5181287ab5553cda0333aabfac50e623f28/integration_tests/enum_01.py#L15-L19

ASR owner of `color` variable is the function, `test_selected_color` and not `Color` enumeration. I can distinguish between the two in the backend i.e., when the variable is of Enum type then extract EnumType from it otherwise do `symbol_symtab(v)->asr_owner` but then it would be boilerplate for the backend which could be easily avoided by storing `enum_type` in the ASR node itself. Let me know what you think.

Reply:
It seems there are two use cases `EnumName(color, ...)` and `EnumName(Color.RED, ...)`. The `enum_type` should be `Color` in both, correct?

Isn't `enum_type` just equal to `symbol_type(v)`? If not, there must be a way to get to it, just try to figure it out.

Reply:
As you have pointed there are two use cases,

1. `EnumName(color, ...)` - Here `enum_type = symbol_type(color)`, so not a problem.
2. `EnumName(Color.RED, ...)` - Here `enum_type != symbol_type(Color.RED)`, because `Color.RED` is of `i32` type (the common member type of an Enum). `Color.RED` is a `Variable` which has `m_parent_symtab` which is the scope of `Enum`. But `Variable.m_parent_symtab->asr_owner` should be the `EnumType`.

So clearly you will have to do,

```
if( ASR::is_a<ASR::Enum_t>(*ASRUtils::symbol_type(x.m_v) ) {
    // Extract EnumType from Enum_t.
} else {
   // Do Variable.m_parent_symtab->asr_owner
}
```

The above if-else check will happen whenever you want to access the global enum values and name array. That's why I think we shouldn't use this if-else check but instead store it in `EnumName` and `EnumValue` node as `ttype enum_type`. What do you think? Do you want me to use the above if-else check instead?

Reply:
Opened https://github.com/lcompilers/lpython/pull/1133. Seems like the if-else check is simple enough, encapsulated it in a function and called it from backends. So I am going to merge the PR anyways.

Reply:
I see. It's good enough for now. I am still not sure about this, it seems there is a better way, but it's good enough for now.